File: MoxieChat/ViewController.m
//
//  ViewController.m
//  MoxieChat
//
//  Created by Moxtra on 2017/6/26.
//  Copyright © 2017年 Moxtra. All rights reserved.
//

#import "ViewController.h"
#import <ChatSDK/MXChatSDK.h>


/*SDK configuration */
#define MOXTRASDK_BASE_DOMAIN @"www.grouphour.com"
#define MOXTRASDK_CLIENT_ID @"ODMyNGYyMzI"
#define MOXTRASDK_CLIENT_SECRET @"OWI5OGMzMjM"

#define MOXTRASDK_UNIQUE_ID  @"mc_jacob"
#define MOXTRASDK_ORG_ID  @"PP2iXx2qi4f46aic2WD01Y7"

@interface ViewController ()<UITableViewDelegate, UITableViewDataSource, MXChatClientDelegate, MXChatListModelDelegate>

@property (nonatomic, strong) UITableView *tableView;

@property (nonatomic, strong) MXChatListModel *chatListModel;
@property (nonatomic, strong) MXMeetListModel *meetListModel;

@property (nonatomic, strong) NSMutableArray<MXChat *> *chatList;
@property (nonatomic, strong) MXChat *lastOpenedChatItem;

@end

static NSString *const kCellReustIdentifier = @"kCellReustIdentifier";

@implementation ViewController

#pragma mark - LifeCycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    [MXChatClient sharedInstance].delegate = self;
    [self login];
    
    [self setupNavigationItem];

    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [self.view addSubview:self.tableView];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSMutableArray<MXChat *> *)chatList
{
    if (_chatList == nil)
    {
        _chatList = [[NSMutableArray alloc] init];
    }
    return _chatList;
}

- (MXChatListModel *)chatListModel
{
    if (_chatListModel == nil)
    {
        _chatListModel = [[MXChatListModel alloc] init];
        _chatListModel.delegate = self;
    }
    return _chatListModel;
}

- (MXMeetListModel *)meetListModel
{
    if (_meetListModel == nil)
    {
        _meetListModel = [[MXMeetListModel alloc] init];
    }
    return _meetListModel;
}

#pragma mark - UserInterface

- (void)setupNavigationItem
{
    self.navigationItem.leftBarButtonItems = [NSArray arrayWithObjects:[[UIBarButtonItem alloc] initWithTitle:@"APIs" style:UIBarButtonItemStylePlain target:self action:@selector(apiButtonOnClicked:)],
                                              nil];
}

#pragma mark - UITableViewDataSource/Delegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.chatList.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellReustIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:kCellReustIdentifier];
    }
    
    MXChat *chatItem = [self.chatList objectAtIndex:indexPath.row];
    cell.textLabel.text = [NSString stringWithFormat:@"%@",chatItem.topic];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    MXChat *chatItem = [self.chatList objectAtIndex:indexPath.row];
    MXChatViewController *chatViewController = [[MXChatViewController alloc] initWithChat:chatItem];
    chatViewController.edgesForExtendedLayout = UIRectEdgeNone;
    self.lastOpenedChatItem = chatItem;
    [self.navigationController pushViewController:chatViewController animated:YES];
}

#pragma mark - WidgetsAction

- (void)apiButtonOnClicked:(UINavigationItem *)sender
{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    __weak typeof(self) weakSelf = self;
    [alert addAction:[UIAlertAction actionWithTitle:@"Login" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf login];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Unlink Account" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf unLink];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Fetch Chat List" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf fetchChatList];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"New Chat" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf startNewChat];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Start Meet" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf startMeet];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Join Meet" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf joinMeet];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Delete Last Open Chat" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf deleteLastOpenChat];
    }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        
    }]];
    alert.popoverPresentationController.barButtonItem = self.navigationItem.leftBarButtonItem;
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - SDKActions

- (void)login
{
    __weak typeof(self) weakSelf = self;
    [[MXChatClient sharedInstance] setupDomain:MOXTRASDK_BASE_DOMAIN httpsDomain:nil wssDomain:nil linkConfig:nil];
    [[MXChatClient sharedInstance] linkWithUniqueId:MOXTRASDK_UNIQUE_ID orgId:MOXTRASDK_ORG_ID clientId:MOXTRASDK_CLIENT_ID clientSecret:MOXTRASDK_CLIENT_SECRET completionHandler:^(NSError * _Nullable error) {
            [weakSelf alertWithMessage:@"Login success"];
            [weakSelf fetchChatList];
    }];
}

- (void)unLink
{
    [[MXChatClient sharedInstance] unlink];
    [self.chatList removeAllObjects];
    [self.tableView reloadData];
}

- (void)fetchChatList
{
    if ([self checkLogin])
    {
        self.chatList = [self.chatListModel.chats mutableCopy];
        [self.tableView reloadData];
    }
}

- (void)startNewChat
{
    if ([self checkLogin])
    {
        __block UITextField *alertTextFiled;
        __weak typeof(self) weakSelf = self;
        UIAlertController *addChatAlert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Chat's Topic", @"Chat's Topic") message:nil preferredStyle:UIAlertControllerStyleAlert];
        [addChatAlert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
            alertTextFiled = textField;
        }];
        
        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel") style:UIAlertActionStyleCancel handler:nil];
        UIAlertAction *confirmAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"Done", @"Done") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [weakSelf.chatListModel createChatWithTopic:alertTextFiled.text completionHandler:^(NSError * _Nullable error, MXChat * _Nullable chatItem) {
                if (error)
                {
                    [weakSelf alertWithError:error];
                }
                else
                {
                    MXChatViewController *chatViewController = [[MXChatViewController alloc] initWithChat:chatItem];
                    
                    chatViewController.edgesForExtendedLayout = UIRectEdgeNone;
                    weakSelf.lastOpenedChatItem = chatItem;
                    [weakSelf.navigationController pushViewController:chatViewController animated:YES];
                }
            }];
        }];
        [addChatAlert addAction:cancelAction];
        [addChatAlert addAction:confirmAction];
        [self presentViewController:addChatAlert animated:YES completion:nil];
    }
}

- (void)startMeet
{
    if ([self checkLogin])
    {
        [self.meetListModel startMeetWithTopic:@"Meet" completionHandler:^(NSError * _Nullable error, MXMeetSession * _Nullable meetSession) {
            [meetSession presentMeetWindow];
        }];
    }
}

- (void)joinMeet
{
    if ([self checkLogin])
    {
        __weak typeof(self) weakSelf = self;
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Input Meet ID" message:nil preferredStyle:UIAlertControllerStyleAlert];
        
        [alert addTextFieldWithConfigurationHandler:^(UITextField *textField) {
            textField.placeholder = @"meet's id";
        }];
        
        [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            
        }]];
        
        [alert addAction:[UIAlertAction actionWithTitle:@"Join" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            
            NSArray * textfields = alert.textFields;
            UITextField * idfield = textfields[0];
            [self.meetListModel joinMeetWithMeetId:idfield.text completionHandler:^(NSError * _Nullable error, MXMeetSession * _Nullable meetSession) {
                if (error)
                {
                    [weakSelf alertWithError:error];
                }
                else
                {
                    [meetSession presentMeetWindow];
                }
            }];
            
        }]];
        
        [self presentViewController:alert animated:YES completion:nil];
    }
}

- (void)deleteLastOpenChat
{
    if ([self checkLogin])
    {
        if (self.lastOpenedChatItem != nil)
        {
            __weak typeof(self) weakSelf = self;
            [self.chatListModel deleteOrLeaveChat:self.lastOpenedChatItem withCompletionHandler:^(NSError * _Nullable error) {
                if (error)
                {
                    [weakSelf alertWithError:error];
                }
                else
                {
                    weakSelf.lastOpenedChatItem = nil;
                }
            }];
        }
        else
        {
            [self alertWithMessage:@"Could not found last open chat"];
        }
    }
}

#pragma mark - Helper

- (void)alertWithError:(NSError *)error
{
    NSString *title = @"Error";
    NSString *message = [NSString stringWithFormat:@"domain:%@, code:%@\n%@", error.domain, @(error.code), error.localizedDescription];
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)alertWithMessage:(NSString *)message
{
    NSString *title = @"Tip";
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil];
}

- (BOOL)checkLogin
{
    if ([MXChatClient sharedInstance].currentUser != nil)
    {
        return YES;
    }
    else
    {
        [self alertWithMessage:@"Please login first"];
        return NO;
    }
}

#pragma mark - MXChatClientDelegate

- (void)chatClientDidUnlink:(MXChatClient *)chatClient
{
    self.chatListModel = nil;
    self.meetListModel = nil;
    [self alertWithMessage:@"Unlink success"];
}

#pragma mark - MXChatListModel

- (void)chatListModel:(MXChatListModel *)chatListModel didCreateChats:(NSArray<MXChat *> *)createdChats
{
    if (createdChats)
    {
        [self.chatList insertObjects:createdChats atIndexes:[NSIndexSet indexSetWithIndex:0]];
        [self.tableView reloadData];
    }
}

- (void)chatListModel:(MXChatListModel *)chatListModel didUpdateChats:(NSArray<MXChat *> *)updatedChats
{
    [self.tableView reloadData];
}

- (void)chatListModel:(MXChatListModel *)chatListModel didDeleteChats:(NSArray<MXChat *> *)deletedChats
{
    if (deletedChats)
    {
        [self.chatList removeObjectsInArray:deletedChats];
        [self.tableView reloadData];
    }
}

@end

File: SwiftMoxie/SwiftMoxie/ViewController.swift
//
//  ViewController.swift
//  SwiftMoxie
//
//  Created by Moxtra on 2017/7/24.
//  Copyright © 2017年 moxtra. All rights reserved.
//

import UIKit

/*SDK configuration */
private let kBaseDomain = "www.grouphour.com"
private let kCLientID = "ODMyNGYyMzI"
private let kClientSecret = "OWI5OGMzMjM"
private let kUniqueID = "mc_jacob"
private let kOrgID = "PP2iXx2qi4f46aic2WD01Y7"

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource,MXChatClientDelegate,MXChatListModelDelegate {
    private let kCellReuseIdentifier = "kCellReustIdentifier"
    private var tableView:UITableView!
    
    private var chatListModel:MXChatListModel {
        get {
            if _chatListModel == nil {
                _chatListModel = MXChatListModel.init()
                _chatListModel!.delegate = self
            }
            return _chatListModel!
        }
    }
    var _chatListModel:MXChatListModel?
    
    private var meetListModel:MXMeetListModel {
        if _meetListModel == nil {
            _meetListModel = MXMeetListModel.init()
        }
        return _meetListModel!
    }
    var _meetListModel:MXMeetListModel?
    
    private lazy var chatList:Array<MXChat> = {
       return Array.init()
    }()
    
    private var lastOpenChat:MXChat!
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    
    override func viewDidLoad() {
        super.viewDidLoad()
        MXChatClient.sharedInstance().delegate = self
        login()
        setupUserInterface()
    }
    
    //MARK: - UserInterface
    private func setupUserInterface() {
        tableView = UITableView.init(frame: view.bounds, style: .plain)
        tableView.delegate = self
        tableView.dataSource = self
        tableView.autoresizingMask = [.flexibleHeight, .flexibleWidth]
        view.addSubview(tableView)
        
        let leftBarButtonItem = UIBarButtonItem.init(title: "APIs", style: .plain, target: self, action: #selector(apiButtonOnClicked))
        navigationItem.leftBarButtonItem = leftBarButtonItem

    }
    
    //MARK: - TableViewDelegate/DataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return chatList.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCell(withIdentifier: kCellReuseIdentifier)
        if cell == nil {
            cell = UITableViewCell.init(style: .default, reuseIdentifier: kCellReuseIdentifier)
        }
        let chatItem = chatList[indexPath.row]
        cell?.textLabel?.text = chatItem.topic
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let chatItem = chatList[indexPath.row]
        let chatViewController = MXChatViewController.init(chat: chatItem)
        chatViewController.edgesForExtendedLayout = UIRectEdge.init(rawValue: 0)
        self.lastOpenChat = chatItem;
        navigationController?.pushViewController(chatViewController, animated: true)
    }
    
    //MARK: - WidgetsAction
    @objc private func apiButtonOnClicked(sender: UIBarButtonItem) {
        weak var weakSelf = self
        
        let alert = UIAlertController.init(title: nil, message: nil, preferredStyle: .actionSheet)
        alert.addAction(UIAlertAction.init(title: "Login", style: .default, handler: { (action) in
            weakSelf?.login()
        }))
        
        alert.addAction(UIAlertAction.init(title: "Unlink Accout", style: .default, handler: { (action) in
            weakSelf?.unlink()
        }))
        
        alert.addAction(UIAlertAction.init(title: "Fetch Chat List", style: .default, handler: { (action) in
            weakSelf?.fetchChatList()
        }))
        
        alert.addAction(UIAlertAction.init(title: "New Chat", style: .default, handler: { (action) in
            weakSelf?.startNewChat()
        }))
        
        alert.addAction(UIAlertAction.init(title: "Start Meet", style: .default, handler: { (action) in
            weakSelf?.startMeet()
        }))
        
        alert.addAction(UIAlertAction.init(title: "Join Meet", style: .default, handler: { (action) in
            weakSelf?.joinMeet()
        }))
        
        alert.addAction(UIAlertAction.init(title: "Delete Last Open Chat", style: .default, handler: { (action) in
            weakSelf?.deleteLastOpenChat()
        }))
        
        alert.addAction(UIAlertAction.init(title: "Cancel", style: .cancel, handler: { (action) in
            
        }))
        
        alert.popoverPresentationController?.barButtonItem = sender
        present(alert, animated: true, completion: nil)
    }
    
    
    //MARK: - ChatSDKActions
    private func login() {
        weak var weakSelf = self
        MXChatClient.sharedInstance().setupDomain(kBaseDomain, httpsDomain: nil, wssDomain: nil, linkConfig: nil)
        MXChatClient.sharedInstance().link(withUniqueId: kUniqueID, orgId: kOrgID, clientId: kCLientID, clientSecret: kClientSecret) { (error) in
            if (error != nil) {
                weakSelf?.alertWithError(error: error!)
            }
            else {
                weakSelf?.alertWithMessage(message: "Login Success")
            }
        }
    }
    
    private func unlink() {
        MXChatClient.sharedInstance().unlink()
        chatList.removeAll()
        tableView.reloadData()
    }
    
    private func fetchChatList() {
        if checkLogin() {
            chatList = chatListModel.chats()
            tableView.reloadData()
        }
    }
    
    private func startNewChat() {
        if checkLogin() {
            weak var weakSelf = self
            weak var blockTextFiled:UITextField?
            let addChatAlert = UIAlertController.init(title: "Chat's Topic", message: nil, preferredStyle: .alert)
            addChatAlert.addTextField { (textField) in
                blockTextFiled = textField
            }
            addChatAlert.addAction(UIAlertAction.init(title: "Cancel", style: .cancel, handler: nil))
            addChatAlert.addAction(UIAlertAction.init(title: "Done", style: .default, handler: { (action) in
                weakSelf?.chatListModel.createChat(withTopic: blockTextFiled?.text, completionHandler: { (error, chatItem) in
                    if (error != nil) {
                        weakSelf?.alertWithError(error: error!)
                    } else {
                        let chatViewController = MXChatViewController.init(chat: chatItem!)
                        chatViewController.edgesForExtendedLayout = UIRectEdge.init(rawValue: 0)
                        weakSelf?.lastOpenChat = chatItem
                        weakSelf?.navigationController?.pushViewController(chatViewController, animated: true)
                    }
                })
            }))
            present(addChatAlert, animated: true, completion: nil)
        }
    }
    
    private func startMeet() {
        if checkLogin() {
            weak var weakSelf = self
            meetListModel.startMeet(withTopic: "New Meet", completionHandler: { (error, meetSession) in
                if (error != nil) {
                    weakSelf?.alertWithError(error: error!)
                } else {
                    meetSession?.presentMeetWindow()
                }
            })
        }
    }
    
    private func joinMeet() {
        if checkLogin() {
            weak var weakSelf = self
            weak var blockTextField:UITextField?
            let joinMeetAlert = UIAlertController.init(title: "Input Meet ID", message: nil
                , preferredStyle: .alert)
            joinMeetAlert.addTextField { (textField) in
                textField.placeholder = "meet's id"
                blockTextField = textField
            }
            joinMeetAlert.addAction(UIAlertAction.init(title: "Cancel", style: .cancel, handler: nil))
            joinMeetAlert.addAction(UIAlertAction.init(title: "Join", style: .default, handler: { (action) in
                weakSelf?.meetListModel.joinMeet(withMeetId: (blockTextField?.text)!, completionHandler: { (error, meetSession) in
                    if error != nil {
                        weakSelf?.alertWithError(error: error!)
                    } else {
                        meetSession?.presentMeetWindow()
                    }
                })
            }))
            present(joinMeetAlert, animated: true, completion: nil)
        }
    }
    
    private func deleteLastOpenChat() {
        if checkLogin() {
            weak var weakSelf = self
            if lastOpenChat != nil {
                chatListModel.deleteOrLeaveChat(lastOpenChat, withCompletionHandler: { (error) in
                    if error != nil {
                        weakSelf?.alertWithError(error: error!)
                    } else {
                        weakSelf?.lastOpenChat = nil
                    }
                })
            }
            else {
                alertWithMessage(message: "Could not found last open chat")
            }
        }
    }
    
    //MARK: - Helper
    private func alertWithError(error: Error) {
        let title = "Error"
        let message = String.init(format: "%@", error.localizedDescription)
        let alert = UIAlertController.init(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction.init(title: "OK", style: .default, handler:nil))
        present(alert, animated: true, completion: nil)
    }
    
    private func alertWithMessage(message: String) {
        let title = "Tip"
        let alert = UIAlertController.init(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction.init(title: "OK", style: .default, handler:nil))
        present(alert, animated: true, completion: nil)
    }
    
    private func checkLogin() -> Bool {
        if  MXChatClient.sharedInstance().currentUser != nil {
            return true
        } else {
            alertWithMessage(message: "Please Login first")
            return false
        }
    }
    
    
    //MARK: - MXChatClientDelegate
    func chatClientDidUnlink(_ chatClient: MXChatClient) {
        _chatListModel = nil
        _meetListModel = nil
        lastOpenChat = nil
        alertWithMessage(message: "Unlink success")
    }
    
    //MARK: - MXChatListModelDelegate
    func chatListModel(_ chatListModel: MXChatListModel, didCreateChats createdChats: [MXChat]) {
        for chatItem in createdChats {
            chatList.insert(chatItem, at: 0)
        }
        tableView.reloadData()
    }
    
    func chatListModel(_ chatListModel: MXChatListModel, didUpdate updatedChats: [MXChat]) {
        tableView.reloadData()
    }

    func chatListModel(_ chatListModel: MXChatListModel, didDelete deletedChats: [MXChat]) {
        for chatItem in deletedChats {
            chatList.remove(at: chatList.index(of: chatItem)!)
        }
        tableView.reloadData()
    }
}

